From 041982858e31aef6c631953ad7b2c959c94190e2 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Sat, 19 Jul 2014 04:30:11 -0400
Subject: [pci] msi: Check kmalloc() return value, fix leak of name

Message-id: <20140719043010.19438.33715.stgit@gir.stowe>
Patchwork-id: 86033
O-Subject: [RHEL7.1 PATCH 71/81] PCI/MSI: Check kmalloc() return value, fix leak of name
Bugzilla: 1110895
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110895
Upstream Status: 86bb4f697a964ab5e50f1159e04c6a5d5acde2d4

commit 86bb4f697a964ab5e50f1159e04c6a5d5acde2d4
Author: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date:   Thu Feb 13 10:47:20 2014 -0700

    PCI/MSI: Check kmalloc() return value, fix leak of name

    Coverity reported that I forgot to check the return value of kmalloc() when
    creating the MSI attribute name, so fix that up and properly free it if
    there is an error when allocating the msi_dev_attr variable.

    Found by Coverity (CID 1163315 and 1163316).

    Fixes: 1c51b50c2995 ("PCI/MSI: Export MSI mode using attributes, not kobject
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/pci/msi.c b/drivers/pci/msi.c
index 6b00032..6c423f9 100644
--- a/drivers/pci/msi.c
+++ b/drivers/pci/msi.c
@@ -567,9 +567,15 @@ static int populate_msi_sysfs(struct pci_dev *pdev)
   return -ENOMEM;
  list_for_each_entry(entry, &pdev->msi_list, list) {
   char *name = kmalloc(20, GFP_KERNEL);
+  if (!name)
+   goto error_attrs;
+
   msi_dev_attr = kzalloc(sizeof(*msi_dev_attr), GFP_KERNEL);
-  if (!msi_dev_attr)
+  if (!msi_dev_attr) {
+   kfree(name);
    goto error_attrs;
+  }
+
   sprintf(name, "%d", entry->irq);
   sysfs_attr_init(&msi_dev_attr->attr);
   msi_dev_attr->attr.name = name;
-- 
1.7.1